### ISSUE TITLE
Command changeServiceForVirtualMachine fails to update read/write rate

### ISSUE THREAD

**Issue by GabrielBrascher** [apache/cloudstack#4125]
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
API
~~~

##### CLOUDSTACK VERSION

Issue was reproduced on (probably it happens also on 4.13.0,  4.13.1):
~~~
4.12.0.0,
4.14.0.0,
master (4.15.0.0)
~~~

##### OS / ENVIRONMENT
VMs are running on **KVM** nodes.

##### SUMMARY

When changing the service offering of a VM from an offering that has **bytes_read_rate_max** configured, the VM does not update **bytes_read_rate_max**.

For instance, if the new offering has **bytes_read_rate** higher than the previous offering **bytes_read_rate_max** then the VM will fail to start after changing to the new offering.

##### STEPS TO REPRODUCE

Let's assume the follwoing offerings:

```
mysql> select name, type, bytes_write_rate, bytes_write_rate_max from disk_offering where name="OldOffering" or name="NewOffering";
+---------------+---------+------------------+----------------------+
| name          | type    | bytes_write_rate | bytes_write_rate_max |
+---------------+---------+------------------+----------------------+
| OldOffering   | Service |        268435456 |            402653184 |
| NewOffering   | Service |        536870912 |            805306368 |
+---------------+---------+------------------+----------------------+
```

1. Deploy VM with offering **OldOffering**
```
(...)
  <iotune>
    <write_bytes_sec>268435456</write_bytes_sec>
    (...)
    <write_bytes_sec_max>402653184</write_bytes_sec_max>
    (...)
  </iotune>
(...)
```

2. Stopped VM and changed offering to **NewOffering** (via `changeServiceForVirtualMachine` command)

3. VM fails to start with **NewOffering**.

```
2020-06-02 09:28:53,882 WARN [resource.wrapper.LibvirtStartCommandWrapper] (agentRequest-Handler-3:null) (logid:7e3f6a86) LibvirtException 
org.libvirt.LibvirtException: internal error: process exited while connecting to monitor: 2020-06-02T12:28:53.743449Z qemu-system-x86_64: -drive file=/mnt/37c2fbd0-8518-3c47-872d-1a1bc772372a/93164fa0-d5ae-417f-8e23-10303163c8c3,format=qcow2,if=none,id=drive-ide0-0-0,ser
ial=93164fa0d5ae417f8e23,cache=none,throttling.bps-read=536870912,throttling.bps-write=536870912,throttling.iops-read=2000,throttling.iops-write=1000,throttling.bps-read-max=805306368,throttling.bps-write-max=402653184: 'serial' is deprecated, please use the correspondin
g option of '-device' instead 
2020-06-02T12:28:53.743499Z qemu-system-x86_64: -drive file=/mnt/37c2fbd0-8518-3c47-872d-1a1bc772372a/93164fa0-d5ae-417f-8e23-10303163c8c3,format=qcow2,if=none,id=drive-ide0-0-0,serial=93164fa0d5ae417f8e23,cache=none,throttling.bps-read=536870912,throttling.bps-write=536
870912,throttling.iops-read=2000,throttling.iops-write=1000,throttling.bps-read-max=805306368,throttling.bps-write-max=402653184: bps_max/iops_max cannot be lower than bps/iops 
       at org.libvirt.ErrorHandler.processError(Unknown Source)
```

Note that `throttling.bps-write=536870912` is higher than `throttling.bps-write-max=402653184`, which results on the above exception (`bps_max/iops_max cannot be lower than bps/iops`)

The XML file used by Libvirt holds the following values on `<iotune>` section:

```
(...)
  <iotune>
    <write_bytes_sec>536870912</write_bytes_sec>
    (...)
    <write_bytes_sec_max>402653184</write_bytes_sec_max>
    (...)
  </iotune>
(...)
```

##### EXPECTED RESULTS

1. Deployed VM with offering **OldOffering**
```
(...)
  <iotune>
    <write_bytes_sec>268435456</write_bytes_sec>
    (...)
    <write_bytes_sec_max>402653184</write_bytes_sec_max>
    (...)
  </iotune>
(...)
```

2. Stopped VM and changed offering to **NewOffering**

3. VM starts with **NewOffering** without any problem. XML updates the <iotune rate accordingly to the offering.

```
(...)
  <iotune>
    <write_bytes_sec>536870912</write_bytes_sec>
    (...)
    <write_bytes_sec_max>805306368</write_bytes_sec_max>
    (...)
  </iotune>
(...)
```

##### Observation:
All seems to be working fine when changing the offering via UI (API command `scaleVirtualMachine`). The problem, therefore, is related to the API command `changeServiceForVirtualMachine`.

**Comment by GabrielBrascher** [comment]
Does Dows anyone know if this command is deprecated? API request `scaleVirtualMachine` worked fine and does pretty much the same as `changeServiceForVirtualMachine`.

Any thoughts? @rhtyd @andrijapanicsb @wido @svenvogel @kiwiflyer @DaanHoogland and others.

**Comment by DaanHoogland** [comment]
@GabrielBrascher I *think* it should work as `scaleVirtualMachine` would only work if the offering allows it, would it?

**Comment by rhtyd** [comment]
is this merged and can be closed @GabrielBrascher ?

**Comment by GabrielBrascher** [comment]
@rhtyd @DaanHoogland I am not sure if this issue should be closed. The problem still exists but we applied workarounds by not using this API command.

My proposal then would be to close this issue and start a discussion (or open a new issue) to verify the need to deprecate the command `changeServiceForVirtualMachnie`.

**Comment by DaanHoogland** [comment]
not sure what you are saying, @GabrielBrascher .
1. the mentioned PR does not solve your problem, or the problem is still present in 4.14? Did you check master with #4341 merged?
2. there is an API you think should be deprecated, as opposed to fixed? (/me says fix changeServiceForVirtualMachine or descibe the unexpected behaviour in the API-docs.)

**Comment by GabrielBrascher** [comment]
@DaanHoogland this issue is clear about the problem (that still exists).

Both of the merged PRs that are linked to this issue have nothing to do with **fixing** this API command, but instead, they are making sure to not use `changeServiceForVirtualMachnie`. That is why I mentioned (here https://github.com/apache/cloudstack/issues/4125#issuecomment-729679051) that we should either keep this issue open or then we need to discuss deprecating the problematic command; which might be the best option as the problematic issue is not being used and when it is used it causes problems

#4341 refers to this issue just to mention that the manual tests failed due to using the problematic API command reported here, when changed to `scaleVirtualMachine` then the implementation at PR #4341 works fine.

https://github.com/apache/cloudstack-primate/pull/818 regards to ensuring that the primate UI is not using the problematic API.

**Comment by DaanHoogland** [comment]
ok, best open a mail discuss, I am really against deprecating a API because it is broken for some use, but if we find that it is conceptually wrong, soit.
Let{s keep this open in case someone feels inspired to fix the issue until discussion is closed.

**Comment by GabrielBrascher** [comment]
Just to avoid confusion and highlight the point regarding **deprecating**. The problematic command is buggy, with at least two open issues (#3110, and #4125) describing different issues on the same command.

On the other hand, command `scaleVirtualMachine` does everything that `changeServiceForVirtualMachnie` does (and more), with a difference that one actually works and the other is a problem that we have been hiding for long by not using on UI.

**Comment by slavkap** [comment]
Hi all, I'm not sure exactly how the contribution works - I mean if somebody raise an issue (like this) does he/she should fix it or somebody else could propose a fix outside of the community? I think that I have a solution about the problem, and I've tested on 4.14. I've found that the volumes table isn't updated with the new offerings. So may I I open a PR for this or somebody else is working on it?
